### DISCUSSION TITLE
reduce not working as expected?

### DISCUSSION THREAD

**Discussion by Thompsonmina** [NillionNetwork/discussions#36] [Bugs]
I am trying to do a reduce operation to return if a secret integer is a member of a secret array, conducted a simple test with a literal integer but for some reason it just doesn't work, the program compiles correctly but attempting to execute it doesn't work

```
def nada_main():
    party1 = Party(name="Party1")
    my_array_1 = Array(SecretInteger(Input(name="my_array_1", party=party1)), size=5)
    # my_int1 = SecretInteger(Input(name="my_int1", party=party1))

    @nada_fn
    def add(a: SecretInteger, b: SecretInteger) -> SecretInteger:
        return  (Integer(-1) < a).if_else(b, a)   // the goal is to eventually find an element in the array, but even this standard test doesn't work
  

    addition = my_array_1.reduce(add, Integer(5))

    out2 = Output(addition, "reduce.operation", party1)

    return [out2]
```
    
    
    
    I get this error instead
   ```
  File "/home/mina/code/zill/nillion-hangman/client_hangman/reduce_client.py", line 51, in main
    store_id = await client.store_secrets(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
py_nillion_client.DealerError: storing secrets: dealer unexpected error: query failed: existing particle lookup failed: input distribution calculation failed: malformed indirection: redundant indirection: the source reference already refers to a memory element

```

When I replace the reduce add with the standard 

 ```
def add(a: SecretInteger, b: SecretInteger) -> SecretInteger:
        return a + b
```
this works, What could be the issue?

**Comment by oceans404** [comment]
Thanks for reporting this! 

I created a project with the [nada tool](https://docs.nillion.com/nada) to try running your code, and also saw an error when I tried to generate-test for your original program. 

- Reducing an array works: https://docs.nillion.com/nada-lang-programs#addition-by-reducing-an-array
- Reducing an array and returning Integers in the if_else as part of the reduce works: https://github.com/NillionNetwork/builder-resources/blob/0570cd4292312c2be753b589f7ec2995605dd511/resources/programs/if_else_reduce.py#L13
- Your use case of reducing an array and returning SecretIntegers in the if_else as part of the reduce results in an error. I'll flag this for the team to look into

```
failed to compile program's MIR: malformed indirection: redundant indirection: the source reference already refers to a memory element
```

If the goal is to find an element in an array, this Battleship program (`check_exists` function) might be helpful: https://github.com/Envoy-VC/battleship-x/blob/b763a77f656448233c2ecfdeb3e53ece046f6757/packages/nillion/src/battleship.py#L39

**Comment by Thompsonmina** [comment]
Thanks for the response and thank you for the link, that helps a lot.

**Reply by wwwehr** [reply]
Awesome @Thompsonmina thanks for writing in. If it's not too early, can you try out the suggestion and mark this issue as "answered" if you are happy with it? Also, note that there is a new SDK out today! :tada: 

Best
